File: code/python/10-mcp-server/mcp_agent.py
"""MCP-Enabled Agent — discovers and calls tools from MCP servers.

Architecture:
  MCPAgent        high-level agent: connect, run, close
  ServerRegistry  manages multiple MCP server connections
  ServerConnection  one long-lived connection to one MCP server

The tool namespace pattern is {server_name}__{tool_name}, so two servers
that both expose a "search" tool never collide.

Usage:
    python mcp_agent.py            # runs the built-in demo
"""
from __future__ import annotations

import asyncio
import json
import os
import sys
from contextlib import AsyncExitStack, asynccontextmanager
from dataclasses import dataclass, field
from typing import AsyncIterator

from mcp import ClientSession, StdioServerParameters
from mcp.client.stdio import stdio_client


# ---------------------------------------------------------------------------
# Data types
# ---------------------------------------------------------------------------

@dataclass
class AgentResult:
    """Returned by MCPAgent.run()."""
    answer: str
    messages: list[dict]
    tools_called: list[str] = field(default_factory=list)


@dataclass
class ToolResult:
    """Result from one MCP tool execution."""
    server_name: str
    tool_name: str
    output: str
    is_error: bool = False


# ---------------------------------------------------------------------------
# LLM provider abstraction  (swap in Anthropic, Ollama, etc.)
# ---------------------------------------------------------------------------

class LLMProvider:
    """Base interface — override ``chat`` in each provider subclass."""

    async def chat(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
    ) -> dict:
        raise NotImplementedError


class OpenAIProvider(LLMProvider):
    """OpenAI chat-completions with function calling."""

    def __init__(self, model: str = "gpt-4o", api_key: str | None = None) -> None:
        from openai import AsyncOpenAI  # lazy import
        self.model = model
        self._client = AsyncOpenAI(api_key=api_key or os.environ.get("OPENAI_API_KEY"))

    async def chat(
        self,
        messages: list[dict],
        tools: list[dict] | None = None,
    ) -> dict:
        kwargs: dict = {"model": self.model, "messages": messages}
        if tools:
            kwargs["tools"] = tools

        response = await self._client.chat.completions.create(**kwargs)
        msg = response.choices[0].message

        return {
            "role": "assistant",
            "content": msg.content,
            "tool_calls": [
                {
                    "id": tc.id,
                    "type": "function",
                    "function": {
                        "name": tc.function.name,
                        "arguments": tc.function.arguments,
                    },
                }
                for tc in (msg.tool_calls or [])
            ],
        }


# ---------------------------------------------------------------------------
# ServerConnection  — one long-lived MCP session
# ---------------------------------------------------------------------------

class ServerConnection:
    """Maintains a subprocess + MCP session for a single server."""

    def __init__(self, name: str) -> None:
        self.name = name
        self.session: ClientSession | None = None
        self.tools: list = []          # list[mcp.types.Tool]
        self._exit_stack = AsyncExitStack()
        self.connected = False

    async def connect(self, params: StdioServerParameters) -> None:
        """Start the server subprocess and initialise the MCP session."""
        try:
            read, write = await self._exit_stack.enter_async_context(
                stdio_client(params)
            )
            self.session = await self._exit_stack.enter_async_context(
                ClientSession(read, write)
            )
            await self.session.initialize()
            result = await self.session.list_tools()
            self.tools = result.tools
            self.connected = True
            print(
                f"[registry] Connected to '{self.name}': "
                f"{len(self.tools)} tools discovered",
                file=sys.stderr,
            )
        except Exception as exc:
            self.connected = False
            await self._exit_stack.aclose()
            raise RuntimeError(
                f"Failed to connect to MCP server '{self.name}': {exc}"
            ) from exc

    async def call_tool(self, tool_name: str, arguments: dict) -> ToolResult:
        """Execute *tool_name* on this server and return a ToolResult."""
        if not self.connected or self.session is None:
            return ToolResult(
                server_name=self.name,
                tool_name=tool_name,
                output=json.dumps({"error": f"Server '{self.name}' is not connected."}),
                is_error=True,
            )
        try:
            result = await self.session.call_tool(tool_name, arguments)
            output = "\n".join(
                item.text for item in result.content if hasattr(item, "text")
            )
            return ToolResult(
                server_name=self.name,
                tool_name=tool_name,
                output=output or "(empty response)",
                is_error=bool(result.isError),
            )
        except Exception as exc:
            self.connected = False
            return ToolResult(
                server_name=self.name,
                tool_name=tool_name,
                output=json.dumps({"error": f"Tool execution failed: {exc}"}),
                is_error=True,
            )

    async def close(self) -> None:
        self.connected = False
        await self._exit_stack.aclose()


# ---------------------------------------------------------------------------
# ServerRegistry  — manages multiple ServerConnections
# ---------------------------------------------------------------------------

class ServerRegistry:
    """Connect to multiple MCP servers and route tool calls."""

    def __init__(self) -> None:
        self.servers: dict[str, ServerConnection] = {}
        # Preserved for reconnection
        self._params: dict[str, StdioServerParameters] = {}

    async def connect(
        self, name: str, params: StdioServerParameters
    ) -> list[dict]:
        """Connect to a server; return its tools in OpenAI format."""
        conn = ServerConnection(name)
        await conn.connect(params)
        self.servers[name] = conn
        self._params[name] = params
        return self._format_tools(conn, "openai")

    # ------------------------------------------------------------------
    # Tool format conversion
    # ------------------------------------------------------------------

    def _format_tools(self, conn: ServerConnection, fmt: str) -> list[dict]:
        return [self._to_openai(t, conn.name) if fmt == "openai"
                else self._to_anthropic(t, conn.name)
                for t in conn.tools]

    def _to_openai(self, mcp_tool, server_name: str) -> dict:
        return {
            "type": "function",
            "function": {
                "name": f"{server_name}__{mcp_tool.name}",
                "description": mcp_tool.description or "",
                "parameters": mcp_tool.inputSchema,
            },
        }

    def _to_anthropic(self, mcp_tool, server_name: str) -> dict:
        return {
            "name": f"{server_name}__{mcp_tool.name}",
            "description": mcp_tool.description or "",
            "input_schema": mcp_tool.inputSchema,
        }

    # ------------------------------------------------------------------
    # Public API
    # ------------------------------------------------------------------

    def get_all_tools(self, fmt: str = "openai") -> list[dict]:
        """Return tools from all *connected* servers, namespaced by server."""
        tools: list[dict] = []
        for conn in self.servers.values():
            if conn.connected:
                tools.extend(self._format_tools(conn, fmt))
        return tools

    async def call_tool(
        self, server_name: str, tool_name: str, arguments: dict
    ) -> str:
        """Execute *tool_name* on *server_name* and return result text."""
        conn = self.servers.get(server_name)
        if conn is None:
            return json.dumps({"error": f"Unknown server: {server_name!r}"})
        result = await conn.call_tool(tool_name, arguments)
        return result.output

    async def reconnect(self, server_name: str) -> bool:
        """Reconnect to a server that disconnected. Returns True on success."""
        params = self._params.get(server_name)
        if params is None:
            return False

        old = self.servers.get(server_name)
        if old:
            await old.close()

        try:
            conn = ServerConnection(server_name)
            await conn.connect(params)
            self.servers[server_name] = conn
            print(f"[registry] Reconnected to '{server_name}'", file=sys.stderr)
            return True
        except Exception as exc:
            print(
                f"[registry] Reconnect failed for '{server_name}': {exc}",
                file=sys.stderr,
            )
            return False

    async def disconnect_all(self) -> None:
        """Close every server connection."""
        for conn in self.servers.values():
            await conn.close()
        self.servers.clear()

    def health_check(self) -> dict:
        """Return a connection-status snapshot for all registered servers."""
        return {
            name: {
                "connected": conn.connected,
                "tool_count": len(conn.tools),
                "tools": [t.name for t in conn.tools],
            }
            for name, conn in self.servers.items()
        }


# ---------------------------------------------------------------------------
# MCPAgent  — the main agent
# ---------------------------------------------------------------------------

class MCPAgent:
    """Provider-agnostic agent that uses MCP for tool discovery and execution.

    1. Call ``connect_server()`` for each MCP server to use.
    2. Call ``run(user_input)`` to start the agent loop.
    3. Call ``close()`` when done.
    """

    MAX_ITERATIONS = 10  # Guard against runaway loops

    def __init__(self, llm_provider: LLMProvider) -> None:
        self.llm = llm_provider
        self.server_registry = ServerRegistry()
        self.tools: list[dict] = []  # OpenAI-format tool schemas

    async def connect_server(
        self, name: str, server_params: StdioServerParameters
    ) -> int:
        """Connect to an MCP server and register its tools.

        Returns the number of tools discovered.
        """
        new_tools = await self.server_registry.connect(name, server_params)
        self.tools.extend(new_tools)
        print(
            f"[agent] +{len(new_tools)} tools from '{name}'. "
            f"Total: {len(self.tools)}",
            file=sys.stderr,
        )
        return len(new_tools)

    async def disconnect_server(self, name: str) -> None:
        """Disconnect from *name* and remove its tools from the registry."""
        conn = self.server_registry.servers.get(name)
        if conn:
            await conn.close()
            del self.server_registry.servers[name]
        # Rebuild the tool list without the disconnected server
        self.tools = self.server_registry.get_all_tools("openai")
        print(f"[agent] Disconnected from '{name}'", file=sys.stderr)

    def _convert_mcp_tool_to_openai(self, mcp_tool, server_name: str) -> dict:
        """Convert an MCP tool schema to OpenAI function-calling format.

        Namespaces the tool as ``{server_name}__{tool_name}``.
        """
        return {
            "type": "function",
            "function": {
                "name": f"{server_name}__{mcp_tool.name}",
                "description": mcp_tool.description or "",
                "parameters": mcp_tool.inputSchema,
            },
        }

    def _convert_mcp_tool_to_anthropic(self, mcp_tool, server_name: str) -> dict:
        """Convert an MCP tool schema to Anthropic tool format."""
        return {
            "name": f"{server_name}__{mcp_tool.name}",
            "description": mcp_tool.description or "",
            "input_schema": mcp_tool.inputSchema,
        }

    async def run(self, user_input: str) -> AgentResult:
        """Run the agent loop with all MCP-discovered tools.

        The loop continues until the LLM returns a final answer
        (no tool_calls) or MAX_ITERATIONS is reached.
        """
        messages: list[dict] = [
            {
                "role": "system",
                "content": (
                    "You are a helpful assistant. Use the available tools whenever "
                    "needed to answer the user's question accurately."
                ),
            },
            {"role": "user", "content": user_input},
        ]
        tools_called: list[str] = []

        for _ in range(self.MAX_ITERATIONS):
            response = await self.llm.chat(
                messages=messages,
                tools=self.tools if self.tools else None,
            )
            messages.append(response)

            tool_calls = response.get("tool_calls") or []
            if not tool_calls:
                # No more tool calls — final answer
                return AgentResult(
                    answer=response.get("content") or "",
                    messages=messages,
                    tools_called=tools_called,
                )

            # Execute each tool call via MCP
            for tc in tool_calls:
                full_name = tc["function"]["name"]
                result = await self._execute_mcp_tool(
                    full_name,
                    json.loads(tc["function"]["arguments"]),
                )
                tools_called.append(full_name)
                messages.append({
                    "role": "tool",
                    "tool_call_id": tc["id"],
                    "content": result.output,
                })

        return AgentResult(
            answer="Maximum iterations reached without a final answer.",
            messages=messages,
            tools_called=tools_called,
        )

    async def _execute_mcp_tool(
        self, full_tool_name: str, arguments: dict
    ) -> ToolResult:
        """Parse ``{server_name}__{tool_name}`` and route to the correct server."""
        if "__" not in full_tool_name:
            return ToolResult(
                server_name="unknown",
                tool_name=full_tool_name,
                output=json.dumps({
                    "error": (
                        f"Invalid tool name format: {full_tool_name!r}. "
                        "Expected '{server_name}__{tool_name}'."
                    )
                }),
                is_error=True,
            )

        server_name, tool_name = full_tool_name.split("__", 1)
        print(
            f"[agent] → {server_name}/{tool_name} args={arguments}",
            file=sys.stderr,
        )
        output = await self.server_registry.call_tool(
            server_name, tool_name, arguments
        )
        return ToolResult(
            server_name=server_name,
            tool_name=tool_name,
            output=output,
        )

    async def close(self) -> None:
        """Disconnect from all MCP servers."""
        await self.server_registry.disconnect_all()
        print("[agent] All servers disconnected.", file=sys.stderr)


# ---------------------------------------------------------------------------
# Demo
# ---------------------------------------------------------------------------

async def run_demo() -> None:
    """Connect to the weather server and run several queries."""
    import pathlib

    here = pathlib.Path(__file__).parent
    weather_server = here / "weather_mcp_server" / "server.py"

    if not weather_server.exists():
        print(
            f"[demo] Weather server not found at {weather_server}.\n"
            "Run from code/python/10-mcp-server/.",
            file=sys.stderr,
        )
        return

    print("=" * 60, file=sys.stderr)
    print("MCP Agent Demo", file=sys.stderr)
    print("=" * 60, file=sys.stderr)

    # Use OpenAIProvider — requires OPENAI_API_KEY env var
    agent = MCPAgent(llm_provider=OpenAIProvider(model="gpt-4o"))

    # Connect to the local weather server
    await agent.connect_server(
        "weather",
        StdioServerParameters(
            command=sys.executable,
            args=[str(weather_server)],
        ),
    )

    # Print discovered tool schemas
    print("\n--- Discovered Tools ---", file=sys.stderr)
    for tool in agent.tools:
        print(f"  {tool['function']['name']}", file=sys.stderr)

    print("\n--- Tool Schemas ---", file=sys.stderr)
    for tool in agent.tools:
        print(
            f"\n{tool['function']['name']}:\n"
            + json.dumps(tool["function"]["parameters"], indent=2),
            file=sys.stderr,
        )

    # Run queries
    queries = [
        "What's the weather in Tokyo?",
        "What's the weather in London and in Dubai?",
        "Give me a 3-day forecast for Paris.",
    ]

    for query in queries:
        print(f"\n{'─' * 50}", file=sys.stderr)
        print(f"Query: {query}", file=sys.stderr)
        result = await agent.run(query)
        print(f"Answer: {result.answer}", file=sys.stderr)
        print(f"Tools called: {result.tools_called}", file=sys.stderr)

    # Health check
    print(
        f"\n--- Health Check ---\n"
        + json.dumps(agent.server_registry.health_check(), indent=2),
        file=sys.stderr,
    )

    # Demonstrate reconnection
    print("\n--- Reconnection Demo ---", file=sys.stderr)
    await agent.disconnect_server("weather")
    success = await agent.server_registry.reconnect("weather")
    print(f"Reconnected: {success}", file=sys.stderr)

    await agent.close()


if __name__ == "__main__":
    asyncio.run(run_demo())

File: code/nodejs/10-mcp-server/mcp_agent.ts
/**
 * MCP-Enabled Agent — TypeScript
 *
 * Discovers and calls tools from MCP servers using the same
 * {serverName}__{toolName} namespace pattern as the Python version.
 *
 * Usage:
 *   OPENAI_API_KEY=sk-... npx tsx mcp_agent.ts
 */

import { Client } from "@modelcontextprotocol/sdk/client/index.js";
import { StdioClientTransport } from "@modelcontextprotocol/sdk/client/stdio.js";
import OpenAI from "openai";
import * as path from "path";
import { fileURLToPath } from "url";

const __filename = fileURLToPath(import.meta.url);
const __dirname  = path.dirname(__filename);

// ---------------------------------------------------------------------------
// Types
// ---------------------------------------------------------------------------

export interface AgentResult {
  answer: string;
  messages: OpenAI.Chat.ChatCompletionMessageParam[];
  toolsCalled: string[];
}

export interface ToolResult {
  serverName: string;
  toolName: string;
  output: string;
  isError: boolean;
}

// Raw MCP tool as returned by client.listTools()
interface McpTool {
  name: string;
  description?: string;
  inputSchema: Record<string, unknown>;
}

// ---------------------------------------------------------------------------
// LLM Provider
// ---------------------------------------------------------------------------

export interface LLMProvider {
  chat(
    messages: OpenAI.Chat.ChatCompletionMessageParam[],
    tools?: OpenAI.Chat.ChatCompletionTool[],
  ): Promise<OpenAI.Chat.ChatCompletionMessage>;
}

export class OpenAIProvider implements LLMProvider {
  private client: OpenAI;

  constructor(
    private model: string = "gpt-4o",
    apiKey?: string,
  ) {
    this.client = new OpenAI({ apiKey: apiKey ?? process.env["OPENAI_API_KEY"] });
  }

  async chat(
    messages: OpenAI.Chat.ChatCompletionMessageParam[],
    tools?: OpenAI.Chat.ChatCompletionTool[],
  ): Promise<OpenAI.Chat.ChatCompletionMessage> {
    const params: OpenAI.Chat.ChatCompletionCreateParamsNonStreaming = {
      model: this.model,
      messages,
      ...(tools && tools.length > 0 ? { tools } : {}),
    };
    const response = await this.client.chat.completions.create(params);
    return response.choices[0]!.message;
  }
}

// ---------------------------------------------------------------------------
// ServerConnection — one long-lived MCP session
// ---------------------------------------------------------------------------

class ServerConnection {
  private client: Client;
  private transport: StdioClientTransport;
  tools: McpTool[] = [];
  connected = false;

  constructor(public readonly name: string) {
    this.client = new Client(
      { name: `agent-client-${name}`, version: "1.0.0" },
      { capabilities: {} },
    );
    // Transport is created during connect()
    this.transport = null as unknown as StdioClientTransport;
  }

  async connect(command: string, args: string[]): Promise<void> {
    this.transport = new StdioClientTransport({ command, args });
    await this.client.connect(this.transport);

    const result = await this.client.listTools();
    this.tools = result.tools as McpTool[];
    this.connected = true;

    process.stderr.write(
      `[registry] Connected to '${this.name}': ${this.tools.length} tools discovered\n`,
    );
  }

  async callTool(toolName: string, toolArgs: Record<string, unknown>): Promise<ToolResult> {
    if (!this.connected) {
      return {
        serverName: this.name,
        toolName,
        output: JSON.stringify({ error: `Server '${this.name}' is not connected.` }),
        isError: true,
      };
    }
    try {
      const result = await this.client.callTool({ name: toolName, arguments: toolArgs });
      const content = result.content as Array<{ type: string; text?: string }>;
      const output = content
        .filter(c => c.type === "text" && c.text)
        .map(c => c.text!)
        .join("\n") || "(empty response)";
      return {
        serverName: this.name,
        toolName,
        output,
        isError: Boolean(result.isError),
      };
    } catch (err) {
      this.connected = false;
      return {
        serverName: this.name,
        toolName,
        output: JSON.stringify({ error: `Tool execution failed: ${err}` }),
        isError: true,
      };
    }
  }

  async close(): Promise<void> {
    this.connected = false;
    await this.transport.close();
  }
}

// ---------------------------------------------------------------------------
// ServerRegistry — manages multiple connections
// ---------------------------------------------------------------------------

interface ConnectParams {
  command: string;
  args: string[];
}

export class ServerRegistry {
  readonly servers: Map<string, ServerConnection> = new Map();
  private params: Map<string, ConnectParams> = new Map();

  async connect(
    name: string,
    command: string,
    args: string[],
  ): Promise<OpenAI.Chat.ChatCompletionTool[]> {
    const conn = new ServerConnection(name);
    await conn.connect(command, args);
    this.servers.set(name, conn);
    this.params.set(name, { command, args });
    return this.formatTools(conn, "openai") as OpenAI.Chat.ChatCompletionTool[];
  }

  private toOpenAI(
    tool: McpTool,
    serverName: string,
  ): OpenAI.Chat.ChatCompletionTool {
    return {
      type: "function",
      function: {
        name: `${serverName}__${tool.name}`,
        description: tool.description ?? "",
        parameters: tool.inputSchema as OpenAI.FunctionParameters,
      },
    };
  }

  private toAnthropic(tool: McpTool, serverName: string): Record<string, unknown> {
    return {
      name: `${serverName}__${tool.name}`,
      description: tool.description ?? "",
      input_schema: tool.inputSchema,
    };
  }

  private formatTools(conn: ServerConnection, fmt: "openai" | "anthropic") {
    return conn.tools.map(t =>
      fmt === "openai" ? this.toOpenAI(t, conn.name) : this.toAnthropic(t, conn.name),
    );
  }

  getAllTools(fmt: "openai" | "anthropic" = "openai") {
    const tools: unknown[] = [];
    for (const conn of this.servers.values()) {
      if (conn.connected) tools.push(...this.formatTools(conn, fmt));
    }
    return tools;
  }

  async callTool(
    serverName: string,
    toolName: string,
    args: Record<string, unknown>,
  ): Promise<string> {
    const conn = this.servers.get(serverName);
    if (!conn) return JSON.stringify({ error: `Unknown server: '${serverName}'` });
    const result = await conn.callTool(toolName, args);
    return result.output;
  }

  async reconnect(serverName: string): Promise<boolean> {
    const p = this.params.get(serverName);
    if (!p) return false;
    const old = this.servers.get(serverName);
    if (old) await old.close();
    try {
      const conn = new ServerConnection(serverName);
      await conn.connect(p.command, p.args);
      this.servers.set(serverName, conn);
      process.stderr.write(`[registry] Reconnected to '${serverName}'\n`);
      return true;
    } catch (err) {
      process.stderr.write(`[registry] Reconnect failed for '${serverName}': ${err}\n`);
      return false;
    }
  }

  async disconnectAll(): Promise<void> {
    for (const conn of this.servers.values()) await conn.close();
    this.servers.clear();
  }

  healthCheck(): Record<string, { connected: boolean; toolCount: number; tools: string[] }> {
    const result: Record<string, { connected: boolean; toolCount: number; tools: string[] }> = {};
    for (const [name, conn] of this.servers) {
      result[name] = {
        connected: conn.connected,
        toolCount: conn.tools.length,
        tools: conn.tools.map(t => t.name),
      };
    }
    return result;
  }
}

// ---------------------------------------------------------------------------
// MCPAgent
// ---------------------------------------------------------------------------

export class MCPAgent {
  readonly serverRegistry: ServerRegistry;
  tools: OpenAI.Chat.ChatCompletionTool[] = [];

  private static readonly MAX_ITERATIONS = 10;

  constructor(private llm: LLMProvider) {
    this.serverRegistry = new ServerRegistry();
  }

  async connectServer(name: string, command: string, args: string[]): Promise<number> {
    const newTools = await this.serverRegistry.connect(name, command, args);
    this.tools.push(...newTools);
    process.stderr.write(
      `[agent] +${newTools.length} tools from '${name}'. Total: ${this.tools.length}\n`,
    );
    return newTools.length;
  }

  async disconnectServer(name: string): Promise<void> {
    const conn = this.serverRegistry.servers.get(name);
    if (conn) {
      await conn.close();
      this.serverRegistry.servers.delete(name);
    }
    this.tools = this.serverRegistry.getAllTools("openai") as OpenAI.Chat.ChatCompletionTool[];
    process.stderr.write(`[agent] Disconnected from '${name}'\n`);
  }

  convertMcpToolToOpenai(
    mcpTool: McpTool,
    serverName: string,
  ): OpenAI.Chat.ChatCompletionTool {
    return {
      type: "function",
      function: {
        name: `${serverName}__${mcpTool.name}`,
        description: mcpTool.description ?? "",
        parameters: mcpTool.inputSchema as OpenAI.FunctionParameters,
      },
    };
  }

  convertMcpToolToAnthropic(
    mcpTool: McpTool,
    serverName: string,
  ): Record<string, unknown> {
    return {
      name: `${serverName}__${mcpTool.name}`,
      description: mcpTool.description ?? "",
      input_schema: mcpTool.inputSchema,
    };
  }

  async run(userInput: string): Promise<AgentResult> {
    const messages: OpenAI.Chat.ChatCompletionMessageParam[] = [
      {
        role: "system",
        content:
          "You are a helpful assistant. Use the available tools whenever needed to answer accurately.",
      },
      { role: "user", content: userInput },
    ];
    const toolsCalled: string[] = [];

    for (let i = 0; i < MCPAgent.MAX_ITERATIONS; i++) {
      const response = await this.llm.chat(
        messages,
        this.tools.length > 0 ? this.tools : undefined,
      );
      messages.push(response as OpenAI.Chat.ChatCompletionMessageParam);

      const toolCalls = response.tool_calls ?? [];
      if (toolCalls.length === 0) {
        return { answer: response.content ?? "", messages, toolsCalled };
      }

      for (const tc of toolCalls) {
        const fullName = tc.function.name;
        const result = await this.executeMcpTool(
          fullName,
          JSON.parse(tc.function.arguments) as Record<string, unknown>,
        );
        toolsCalled.push(fullName);
        messages.push({
          role: "tool",
          tool_call_id: tc.id,
          content: result.output,
        });
      }
    }

    return {
      answer: "Maximum iterations reached without a final answer.",
      messages,
      toolsCalled,
    };
  }

  async executeMcpTool(
    fullToolName: string,
    args: Record<string, unknown>,
  ): Promise<ToolResult> {
    if (!fullToolName.includes("__")) {
      return {
        serverName: "unknown",
        toolName: fullToolName,
        output: JSON.stringify({
          error: `Invalid tool name: '${fullToolName}'. Expected '{server}__{tool}'.`,
        }),
        isError: true,
      };
    }
    const sep = fullToolName.indexOf("__");
    const serverName = fullToolName.slice(0, sep);
    const toolName   = fullToolName.slice(sep + 2);

    process.stderr.write(
      `[agent] → ${serverName}/${toolName} args=${JSON.stringify(args)}\n`,
    );
    const output = await this.serverRegistry.callTool(serverName, toolName, args);
    return { serverName, toolName, output, isError: false };
  }

  async close(): Promise<void> {
    await this.serverRegistry.disconnectAll();
    process.stderr.write("[agent] All servers disconnected.\n");
  }
}

// ---------------------------------------------------------------------------
// Demo
// ---------------------------------------------------------------------------

async function runDemo(): Promise<void> {
  const weatherServer = path.join(
    __dirname,
    "weather_mcp_server",
    "server.ts",
  );

  process.stderr.write("=".repeat(60) + "\n");
  process.stderr.write("MCP Agent Demo (TypeScript)\n");
  process.stderr.write("=".repeat(60) + "\n");

  const agent = new MCPAgent(new OpenAIProvider("gpt-4o"));

  // Connect to weather server (started via tsx)
  await agent.connectServer("weather", "npx", ["tsx", weatherServer]);

  // Show discovered tools
  process.stderr.write("\n--- Discovered Tools ---\n");
  for (const tool of agent.tools) {
    process.stderr.write(`  ${tool.function.name}\n`);
  }

  process.stderr.write("\n--- Tool Schemas ---\n");
  for (const tool of agent.tools) {
    process.stderr.write(
      `\n${tool.function.name}:\n${JSON.stringify(tool.function.parameters, null, 2)}\n`,
    );
  }

  const queries = [
    "What's the weather in Tokyo?",
    "What's the weather in London and Dubai?",
    "Give me a 3-day forecast for Paris.",
  ];

  for (const query of queries) {
    process.stderr.write("\n" + "─".repeat(50) + "\n");
    process.stderr.write(`Query: ${query}\n`);
    const result = await agent.run(query);
    process.stderr.write(`Answer: ${result.answer}\n`);
    process.stderr.write(`Tools called: ${result.toolsCalled.join(", ")}\n`);
  }

  process.stderr.write(
    "\n--- Health Check ---\n" +
    JSON.stringify(agent.serverRegistry.healthCheck(), null, 2) + "\n",
  );

  await agent.close();
}

runDemo().catch(err => {
  process.stderr.write(`[demo] Fatal: ${err}\n`);
  process.exit(1);
});
